File: src/test/java/tech/sirwellington/alchemy/generator/DatesTest.java
package tech.sirwellington.alchemy.generator;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.RepeatedTest;

import java.time.Duration;
import java.time.Instant;
import java.util.Calendar;
import java.util.Date;

import static java.text.MessageFormat.format;
import static java.time.temporal.ChronoUnit.*;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.generator.NumberGenerators.integers;
import static tech.sirwellington.alchemy.generator.NumberGenerators.negativeLongs;
import static tech.sirwellington.alchemy.generator.Throwables.assertThrows;

/**
 * Tests for {@link Dates}.
 *
 * @author SirWellington
 */
@DisplayName("Dates Utilities")
final class DatesTest extends BaseGeneratorTest {

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testNow() {
        // Given
        var result = Dates.now();
        var after = new Date();

        // Then
        assertThat(result, notNullValue());
        assertThat(result.getTime(), lessThanOrEqualTo(after.getTime()));
        assertThat(System.currentTimeMillis() - result.getTime(), lessThan(100L));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testDaysBeforeNow() {
        // Given
        var daysAgo = one(integers(1, 3650));
        var now = Instant.now();
        var expectedLeft = now.minus(Duration.ofDays(daysAgo));

        // When
        var result = Dates.daysBeforeNow(daysAgo);

        // Then
        assertThat(result.getTime(), lessThanOrEqualTo(now.toEpochMilli()));
        assertThat(result.getTime(), greaterThanOrEqualTo(expectedLeft.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testDaysAfterNow() {
        var days = one(integers(1, 3000));
        var now = Instant.now();
        var expectedRight = now.plus(Duration.ofDays(days));

        var result = Dates.daysAfterNow(days);

        assertThat(result.getTime(), greaterThanOrEqualTo(now.toEpochMilli()));
        assertThat(result.getTime(), lessThanOrEqualTo(expectedRight.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testHoursBeforeNow() {
        var hours = one(integers(1, 24 * 365)); // up to ~1 year
        var now = Instant.now();
        var expectedLeft = now.minus(Duration.ofHours(hours));

        var result = Dates.hoursBeforeNow(hours);

        assertThat(result.getTime(), lessThanOrEqualTo(now.toEpochMilli()));
        assertThat(result.getTime(), greaterThanOrEqualTo(expectedLeft.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testHoursAfterNow() {
        // Given
        var hours = one(integers(1, 24 * 365));
        // When
        var result = Dates.hoursAfterNow(hours);
        var now = Instant.now();

        // Then
        var future = now.plus(Duration.ofHours(hours));

        assertThat(result.getTime(), greaterThanOrEqualTo(now.toEpochMilli()));
        assertThat(result.getTime(), lessThanOrEqualTo(future.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testMinutesBeforeNow() {
        var minutes = one(integers(1, 60 * 24)); // up to ~1 day
        var now = Instant.now();
        var expectedLeft = now.minus(Duration.ofMinutes(minutes));

        var result = Dates.minutesBeforeNow(minutes);

        assertThat(result.getTime(), lessThanOrEqualTo(now.toEpochMilli()));
        assertThat(result.getTime(), greaterThanOrEqualTo(expectedLeft.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testMinutesAfterNow() {
        // Given
        int minutes = one(integers(1, 60 * 24));
        // When
        var now = Instant.now();
        var result = Dates.minutesAfterNow(minutes);

        // Then
        var future = now.plus(Duration.ofMinutes(minutes));

        assertThat(result.getTime(), greaterThanOrEqualTo(now.toEpochMilli()));

        if (result.getTime() >= future.toEpochMilli()) {
            System.out.println(
                format("minutes: {0}, result: {1}, future: {2}", minutes, result, future)
            );
        }
        assertThat(result.getTime(), lessThanOrEqualTo(future.toEpochMilli()));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testIsNow_Date() {
        var now = Dates.now();
        assertThat(Dates.isNow(now), is(true));

        Date notNow = Dates.daysBeforeNow(1);
        assertThat(Dates.isNow(notNow), is(false));

        assertThrows(IllegalArgumentException.class, () -> Dates.isNow(null));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testIsNow_Date_long() throws Exception {
        assertThrows(() -> Dates.isNow(new Date(), -1))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> Dates.isNow((Date) null, 0L))
            .isInstanceOf(IllegalArgumentException.class);

        var now = Dates.now();

        assertThat(Dates.isNow(now, 10), is(true));

        Thread.sleep(1);
        assertThat(Dates.isNow(now, 0), is(false));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testIsNow_Instant_long() throws Exception {
        assertThrows(() -> Dates.isNow((Instant) null, 0))
            .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> {
            var now = Instant.now();
            var negativeNumber = one(negativeLongs());
            Dates.isNow(now, negativeNumber);
        }).isInstanceOf(IllegalArgumentException.class);

        var now = Instant.now();
        assertThat(Dates.isNow(now, 10), is(true));

        Thread.sleep(1);
        assertThat(Dates.isNow(now, 0), is(false));
    }

    @RepeatedTest(DEFAULT_ITERATIONS)
    void testIsNow_ToleranceBoundaries() {
        var slightlyEarly = new Date(Instant.now().minusSeconds(2).toEpochMilli());
        var slightlyLate   = new Date(Instant.now().plusSeconds(2).toEpochMilli());
        var oneSecondDuration = Duration.ofSeconds(1).toMillis();
        var fourSecondsDuration = Duration.ofSeconds(4).toMillis();

        assertThat(Dates.isNow(slightlyEarly, fourSecondsDuration), is(true)); // within ±4s
        assertThat(Dates.isNow(slightlyEarly, oneSecondDuration), is(false)); // outside ±1s

        assertThat(Dates.isNow(slightlyLate, fourSecondsDuration), is(true));
        assertThat(Dates.isNow(slightlyLate, oneSecondDuration), is(false));
    }

    @RepeatedTest(20)
    void testCurrentYear() {
        // Given
        var calendar = Calendar.getInstance();
        var expectedYear = calendar.get(Calendar.YEAR);

        // When
        var result = Dates.currentYear();

        // Then
        assertThat(result, equalTo(expectedYear));
    }
}

File: src/test/java/tech/sirwellington/alchemy/generator/DatesTest.kt
/*
 * Copyright 2015 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.generator

import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.greaterThanOrEqualTo
import org.hamcrest.Matchers.lessThanOrEqualTo
import org.junit.After
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.runners.MockitoJUnitRunner
import tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.negativeIntegers
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.positiveIntegers
import tech.sirwellington.alchemy.generator.Throwables.assertThrows
import java.time.Instant
import java.time.Instant.now
import java.time.temporal.ChronoUnit
import java.time.temporal.ChronoUnit.DAYS
import java.time.temporal.ChronoUnit.MINUTES
import java.util.*


/**

 * @author SirWellington
 */
@RunWith(MockitoJUnitRunner::class)
class DatesTest
{

    @Before
    fun setUp()
    {
    }

    @After
    fun tearDown()
    {

    }

    @Test
    fun testNow()
    {
        println("testNow")
        val result = Dates.now()
        val after = Date()
        assertThat(after.time, greaterThanOrEqualTo(result.time))
    }

    @Test
    fun testDaysAgo()
    {
        println("testDaysAgo")

        val days = positiveIntegers().get()

        val right = now().toEpochMilli()
        val left = now().minus(days.toLong(), DAYS).toEpochMilli()

        val result = Dates.daysAgo(days)
        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))
    }

    @Test
    fun testDaysAhead()
    {
        println("testDaysAhead")
        val days = positiveIntegers().get()

        val left = now().toEpochMilli()

        val result = Dates.daysAhead(days)
        val right = now().plus(days.toLong(), ChronoUnit.DAYS).toEpochMilli()

        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))
    }

    @Test
    fun testHoursAgo()
    {
        println("testHoursAgo")
        val hours = positiveIntegers().get()

        val left = now().minus(hours.toLong(), ChronoUnit.HOURS).toEpochMilli()
        val right = now().toEpochMilli()

        val result = Dates.hoursAgo(hours)

        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))
    }

    @Test
    fun testHoursAhead()
    {
        println("testHoursAhead")
        val hours = positiveIntegers().get()

        val left = Instant.now().toEpochMilli()

        val result = Dates.hoursAhead(hours)
        val right = Instant.now().plus(hours.toLong(), ChronoUnit.HOURS).toEpochMilli()

        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))

    }

    @Test
    fun testMinutesAgo()
    {
        println("testMinutesAgo")
        val minutes = positiveIntegers().get()

        val left = now().minus(minutes.toLong(), MINUTES).toEpochMilli()
        val right = now().toEpochMilli()

        val result = Dates.minutesAgo(minutes)

        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))
    }

    @Test
    fun testMinutesAhead()
    {
        println("testMinutesAhead")
        val minutes = positiveIntegers().get()

        val left = now().toEpochMilli()

        val result = Dates.minutesAhead(minutes)
        val right = now().plus(minutes.toLong(), MINUTES).toEpochMilli()

        assertThat(result.time, greaterThanOrEqualTo(left))
        assertThat(result.time, lessThanOrEqualTo(right))
    }

    @Test
    fun testIsNow_Date()
    {
        println("testIsNow_Date")

        val now = Dates.now()
        assertThat(Dates.isNow(now), `is`(true))

        val notNow = Dates.daysAgo(1)
        assertThat(Dates.isNow(notNow), `is`(false))

        assertThrows { Dates.isNow(null) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testIsNow_Date_long()
    {
        println("testIsNow_Date_long")

        assertThrows { Dates.isNow(null as Date?, 0) }
                .isInstanceOf(IllegalArgumentException::class.java)

        val now = Dates.now()
        assertThat(Dates.isNow(now, 10), `is`(true))
        assertThrows { Dates.isNow(now, -1) }
                .isInstanceOf(IllegalArgumentException::class.java)

        Thread.sleep(1)
        assertThat(Dates.isNow(now, 0), `is`(false))

    }

    @Test
    @Throws(Exception::class)
    fun testIsNow_Instant_long()
    {
        println("testIsNow_Instant_long")

        assertThrows { Dates.isNow(null as Instant?, 0) }
                .isInstanceOf(IllegalArgumentException::class.java)

        assertThrows { Dates.isNow(Instant.now(), one(negativeIntegers()).toLong()) }
                .isInstanceOf(IllegalArgumentException::class.java)

        val now = Instant.now()
        assertThat(Dates.isNow(now, 10), `is`(true))

        Thread.sleep(1)
        assertThat(Dates.isNow(now, 0), `is`(false))
    }

}
